### ISSUE TITLE
On Balance Volume calculation error

### ISSUE THREAD

**Issue by hcgaron** [bukosabino/ta#41]
EDIT:  I updated my original answer because I think this might be better

Thanks for the hard work on the library.  I noticed a bug in the calculation for On Balance Volume.

In the code:

`    df = pd.DataFrame([close, volume]).transpose()

    df['OBV'] = 0

    c1 = close < close.shift(1)

    c2 = close > close.shift(1)

    if c1.any():

        df.loc[c1, 'OBV'] = - volume

    if c2.any():

        df.loc[c2, 'OBV'] = volume

    obv = df['OBV']

    if fillna:

        obv = obv.replace([np.inf, -np.inf], np.nan).fillna(0)

    return pd.Series(obv, name='obv')`

the lines:

```
if c1.any():

    df.loc[c1, 'OBV'] = - volume

if c2.any():

    df.loc[c2, 'OBV'] = volume`
```

don't seem to allow the OBV value to propagate through time.  It is resetting obv = volume every time the close price is higher than the previous period, and obv = - volume every time the price closes lower.

If you initialize the OBV for period zero to be equal to volume at period zero then you should be able to apply a cumulative sum with `cumsum` .  ex:

`df.loc[0,'OBV'] = df.loc[0,'Volume']`

Then assuming you keep the code you have, you can apply `cumsum` to what we currently have as the OBV column:

`df.loc[:,'OBV']= df['OBV'].cumsum()`

This should keep the cumulative sum as needed with as little effort as possible IMO.

**Comment by Jiraiyah** [comment]
> EDIT: I updated my original answer because I think this might be better
> 
> Thanks for the hard work on the library. I noticed a bug in the calculation for On Balance Volume.
> 
> In the code:
> 
> ` df = pd.DataFrame([close, volume]).transpose()
> 
> ```
> df['OBV'] = 0
> 
> c1 = close < close.shift(1)
> 
> c2 = close > close.shift(1)
> 
> if c1.any():
> 
>     df.loc[c1, 'OBV'] = - volume
> 
> if c2.any():
> 
>     df.loc[c2, 'OBV'] = volume
> 
> obv = df['OBV']
> 
> if fillna:
> 
>     obv = obv.replace([np.inf, -np.inf], np.nan).fillna(0)
> 
> return pd.Series(obv, name='obv')`
> ```
> the lines:
> 
> ```
> if c1.any():
> 
>     df.loc[c1, 'OBV'] = - volume
> 
> if c2.any():
> 
>     df.loc[c2, 'OBV'] = volume`
> ```
> don't seem to allow the OBV value to propagate through time. It is resetting obv = volume every time the close price is higher than the previous period, and obv = - volume every time the price closes lower.
> 
> If you initialize the OBV for period zero to be equal to volume at period zero then you should be able to apply a cumulative sum with `cumsum` . ex:
> 
> `df.loc[0,'OBV'] = df.loc[0,'Volume']`
> 
> Then assuming you keep the code you have, you can apply `cumsum` to what we currently have as the OBV column:
> 
> `df.loc[:,'OBV']= df['OBV'].cumsum()`
> 
> This should keep the cumulative sum as needed with as little effort as possible IMO.

I agree, obv should be cumulative sum, until a correct implementation, when should that line you mentioned (cumsum) happen? I assume before 
obv = df['OBV']
an I right?

**Comment by bukosabino** [comment]
Hi @Jiraiyah @hcgaron ,

You are right, OBV need `cumsum` to be calculated. Now is fixed in last version (v0.4.4).

Thank you.
